packages/roosterjs-content-model/lib/editor/ContentModelEditor.ts

```diff
@@ -57,4 +60,51 @@ export default class ContentModelEditor
             core.cachedModel = model || undefined;
         }
     }
+
+    /**
+     * Paste into editor using a clipboardData object
+     * @param clipboardData Clipboard data retrieved from clipboard
+     * @param pasteAsText Force pasting as plain text. Default value is false
+     * @param applyCurrentStyle True if apply format of current selection to the pasted content,
+     * false to keep original format.  Default value is false. When pasteAsText is true, this parameter is ignored
+     */
+    public paste(
+        clipboardData: ClipboardData,
+        pasteAsText: boolean = false,
+        applyCurrentFormat: boolean = false,
+        pasteAsImage: boolean = false
+    ) {
+        const core = this.getCore();
+        if (!clipboardData) {
+            return;
+        }
+
+        if (clipboardData.snapshotBeforePaste) {
+            // Restore original content before paste a new one
+            this.setContent(clipboardData.snapshotBeforePaste);
+        } else {
+            clipboardData.snapshotBeforePaste = this.getContent(
+                GetContentMode.RawHTMLWithSelection
+            );
+        }
+
+        const range = this.getSelectionRange();
+        const pos = range && Position.getStart(range);
+        const model = core.api.createPasteModel(
+            core,
+            clipboardData,
+            pos,
+            pasteAsText,
+            applyCurrentFormat,
+            pasteAsImage
+        );
+
+        if (model) {
+            const currentModel = this.createContentModel();
+
+            mergeModel(currentModel, model);
+
+            this.setContentModel(currentModel);
```

Comment:
Use formatWithContentModel(), and pass in correct change source. Otherwise it won't add undo snapshot.